File: 01_simple_examples/R/02_split_conformal_libraries.R
#################################################
# Split Conformal - Regression
# Libraries
#################################################

# Data Generation
set.seed(1)
n <- 100  
x <- runif(n, 0, 5)
y <- x^2 + rnorm(n, sd = 1.5)

# New Test Instance 
x_new <- 2.5
y_true <- x_new^2 + rnorm(1, sd = 1.5)

alpha <- 0.10


#################################################
# ConformalInference
#################################################

library(conformalInference)
library(randomForest)

# Define Wrapper Functions for the Base Learner
train.fun <- function(x, y) {
    df <- data.frame(x = x[, 1], y = y)
    randomForest(y ~ x, data = df)
}

predict.fun <- function(out, newx) {
    df <- data.frame(x = newx[, 1])
    predict(out, newdata = df)
}

# Run Split Conformal Prediction using the framework
start_time <- Sys.time()

out <- conformal.pred.split(
    x = x,
    y = y,
    x0 = 2.5,
    train.fun = train.fun,
    predict.fun = predict.fun,
    alpha = alpha,
    rho = 2/3
)

end_time <- Sys.time()

execution_time <- as.numeric(end_time - start_time, units = "secs")
print(paste("Execution time:", execution_time, "seconds"))

prediction_interval <- c(out$lo, out$up)
print("ConformalInference Library Interval:")
print(continuous_interval)

#################################################
# Probably
#################################################

library(tidymodels)
library(probably)

start_time <- Sys.time()

# Split dataset into training and calibration sets
idx <- sample(1:n, floor(2/3 * n))

train <- data.frame(x = x[idx], y = y[idx])
cal <- data.frame(x = x[-idx], y = y[-idx])

# Random forest model (tidymodels)
rf_spec <- rand_forest() %>%
    set_engine("ranger") %>%
    set_mode("regression")

# Tidymodels workflow
wf <- workflow() %>%
    add_formula(y ~ x) %>%
    add_model(rf_spec)

fit_model <- fit(wf, data = train)

# Split conformal
conf <- int_conformal_split(
    fit_model,
    cal
)

end_time <- Sys.time()
execution_time <- as.numeric(end_time - start_time, units = "secs")
print(paste("Execution time:", execution_time, "seconds"))

# Prediction + interval
print("Probably Library Interval:")
predict(conf, new_data = data.frame(x = x_new), level = 1 - alpha)

#################################################
# Predictset
#################################################

library(predictset)

start_time <- Sys.time()

# Model and wrapper functions specification
rf_model <- make_model(
    train_fun = function(x, y) {
        randomForest(x = data.frame(x = x), y = y)
    },
    predict_fun = function(object, x_new) {
        predict(object, newdata = data.frame(x = x_new))
    },
    type = "regression"
)

# Calibrate and compute the split conformal prediction
result <- conformal_split(
    x = data.frame(x = x),
    y = y,
    model = rf_model,
    x_new = data.frame(x = x_new),
    alpha = alpha,
    cal_fraction = 0.3
)

end_time <- Sys.time()
execution_time <- as.numeric(end_time - start_time, units = "secs")
print(paste("Execution time:", execution_time, "seconds"))

# Interval
interval <- c(result$lower, result$upper)
print("Predictset Library Interval:")
print(interval)

File: 01_simple_examples/python/02_split_conformal_libraries.py
#################################################
# Split Conformal - Regression
# Libraries
#################################################

import numpy as np
import time
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from mapie.regression import SplitConformalRegressor

# Data Generation
np.random.seed(1)
n = 100
x = np.random.uniform(0, 5, n)
y = x**2 + np.random.normal(0, 1.5, n)

# New Test Instance
x_new = 2.5
y_true = x_new**2 + np.random.normal(0, 1.5)

alpha = 0.10


#################################################
# MAPIE
#################################################

start_time = time.time()

# Split dataset into training and calibration sets
X_train, X_cal, y_train, y_cal = train_test_split(
    x.reshape(-1, 1), y, test_size=1/3, random_state=1
)

# Base model specification
model = RandomForestRegressor(random_state=0)

# Configure the split conformal regressor 
mapie = SplitConformalRegressor(
    estimator=model,
    prefit=False,
    confidence_level=1 - alpha
)

# Model training
mapie.fit(X_train, y_train)

# Calibration step
mapie.conformalize(X_cal, y_cal)

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time:.4f} seconds")

# Prediction and interval generation for the new test instance
y_pred, y_pis = mapie.predict_interval([[x_new]])

# Extract and format the prediction interval
print("MAPIE Prediction Interval:")
print(y_pis)
